app/packs/src/apps/mydb/elements/list/ElementsTable.js

```diff
@@ -231,117 +277,236 @@ export default class ElementsTable extends React.Component {
+
+  renderSamplesHeader = () => {
     const {
-      sampleCollapseAll,
-      moleculeSort, ui,
-      advancedSearch, productOnly
+      moleculeSort,
+      advancedSearch,
+      productOnly,
     } = this.state;
-    const { fromDate, toDate } = ui;
-    const { type, showReport } = this.props;
 
-    const collapseIcon = sampleCollapseAll ? 'chevron-right' : 'chevron-down';
+    const options = [
+      { value: false, label: 'Grouped by Sample' },
+      { value: true, label: 'Grouped by Molecule' }
+    ];
+    const color = productOnly ? '#5cb85c' : 'currentColor';
+
+    return (
+      <>
+        <Select
+          simpleValue
+          options={options}
+          clearable={false}
+          searchable
+          value={moleculeSort}
+          onChange={this.changeSampleSort}
+          className='header-group-select'
+        />
+        <button
+          style={{ border: 'none' }}
+          onClick={this.toggleProductOnly}
+        >
+          <i
+            style={{ cursor: 'pointer', color }}
+            className="fa fa-lg fa-product-hunt"
+          />
+        </button>
+        {this.collapseButton()}
+      </>
+    );
+  }
+
+  renderReactionsHeader = () => {
+    const { elementsGroup, elementsSort } = this.state;
+    const optionsHash = {
+      'none': { sortColumn: 'update date', label: 'List' },
+      'rinchi_short_key': { sortColumn: 'RInChI', label: 'Grouped by RInChI' },
+      'rxno': { sortColumn: 'type', label: 'Grouped by type' },
+    }
+    const options = Object.entries(optionsHash).map((option) => {
+      return { value: option[0], label: option[1].label }
+    });
+    const sortColumn = optionsHash[elementsGroup].sortColumn;
+    const sortTitle = elementsSort ? `sort by ${sortColumn}` : 'sort by update date';
+    const sortTooltip = <Tooltip id="reaction_sort_tooltip">{sortTitle}</Tooltip>;
+    const sortIconClass = elementsSort ? 'fa-sort-alpha-desc' : 'fa-clock-o';
+    const sortIcon = <i className={`fa fa-fw ${sortIconClass}`} />;
+    const sortContent = (
+      <OverlayTrigger placement="top" overlay={sortTooltip}>
+        <button style={{ border: 'none' }} onClick={this.changeElementsSort} >
+          {sortIcon}
+        </button>
+      </OverlayTrigger>
+    );
+
+    return (
+      <>
+        <Select
+          simpleValue
+          options={options}
+          clearable={false}
+          searchable={false}
+          value={elementsGroup}
+          onChange={this.changeElementsGroup}
+          className='header-group-select'
+        />
+        {elementsGroup !== 'none' ? (sortContent) : null}
+        {elementsGroup !== 'none' ? (this.collapseButton()) : null}
+      </>
+    );
+  }
+
+  renderGenericElementsHeader = () => {
+    const { elementsGroup, elementsSort } = this.state;
+    const { genericEl } = this.props;
+
+    let optionsHash = {
+      'none': { sortColumn: 'update date', label: 'List' },
+    }
+    const layers = genericEl.properties_release.layers;
```

Comment:
I've (locally) rebased the `feature/elements-grouping` branch onto the current `main` in order to test the grouping feature.
The app crashed on startup (`bundle exec rails server`) with the following stacktrace:

```
TypeError: Cannot read properties of undefined (reading 'layers')
    at ElementsTable._this.renderGenericElementsHeader (ElementsTable.js:384:1)
    at ElementsTable._this.renderHeader (ElementsTable.js:457:1)
    at ElementsTable.render (ElementsTable.js:573:1)
    at finishClassComponent (react-dom.development.js:17485:1)
    at updateClassComponent (react-dom.development.js:17435:1)
    at beginWork (react-dom.development.js:19073:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
    at invokeGuardedCallback (react-dom.development.js:4056:1)
    at beginWork$1 (react-dom.development.js:23964:1)
```

Reply:
Hey @JanCBrammer sorry for the late reply, 
I tried to reproduce it on my local machine with a new generic element without a release JSON, but it is not throwing any errors. To me, it seems to be an issue after the rebase, and our time budget is all used up for now. Did anything change with how the generic elements now work?

Reply:
Hey @JanCBrammer good day, 
I had some time to work on the grouping story again and pushed a new update.
I adapted some eslinting and rubocop suggestions and rebased with current main.
I also got your error once after a rebase. I could fix that with deleting the content of the klasses.json in config (I don't really understand what exactly it does, but it helped somehow). Later I tried to reproduce it again but failed. The only way I could reproduce it was to delete the properties release content from the element klass, but I am not sure if you can do that without the rails console. In the end, I added a check for the properties release.
I also added some rspec tests for the sorting.

When I started to work on this story, the generic elements didn't work properly, I had to fix some issues. Some of them were the Collectable scopes. They were later fixed differently, but I think my solution is better because it works now for all models, and you don't have duplicated code. I also added a test for the scopes in the samples api spec, and I feel like they don't work like I would expect. I wrote some comments to the concern.

Please tell me if there is anything more I should do to finish this pull request.

Reply:
Thanks for the update @VadimKeller. FYI, @TasnimMehzabin took over the review.